### PR TITLE
Add invert4x4 specialization for invertMatrix()

### PR DESCRIPTION
**Pull request by totto82**
Motivation: Run field case model with CO2 injection.

### CODE REVIEW

**Comment by totto82** [comment]
jenkins build this please

**Comment by totto82** [comment]
jenkins build this ewoms=216 please

**Comment by totto82** [comment]
Visual inspection shows same results for SPE5CASE1. It seems like the number of substeps has changed but comparing the Jenkins run with the reference data in opm-data/spe5 shows no significant difference in convergence. Any idea why this fails?

**Comment by totto82** [comment]
jenkins build this update_data please

**Comment by akva2** [comment]
same problem as yesterday, something must have changed at github :/ opened the pr for you.

**Comment by totto82** [comment]
jenkins build this opm-data=244 please

**Comment by akva2** [comment]
also either you or @GitPaean has to fold, two updates to ref data at the same time aint gonna fly..

File: opm/autodiff/BlackoilModelEbos.hpp [hunk 1]
```diff
@@ -706,6 +706,7 @@ namespace Opm {
                 if (has_solvent_) {
                     double& ss = priVars[Indices::solventSaturationIdx];
                     ss -= satScaleFactor * dss;
+                    ss = std::min(std::max(ss, 0.0),1.0);
```
Comment (blattms) [inline opm/autodiff/BlackoilModelEbos.hpp]: I guess this change is unrelated to this PR and might change the solution?

**Comment by blattms** [comment]
Out of curiosity: Is this needed to stabilize ILU or speed it up?

**Comment by totto82** [comment]
> Out of curiosity: Is this needed to stabilize ILU or speed it up?

Manly to stabilize it. You sometimes get singular or near singular matrices when you use the solvent model. (This is also the case for the legacy solvent model). Returning the identity matrix for these cases  stabilize the run significantly. And is for some of the cases necessary in order to complete the simulations.

**Comment by blattms** [comment]
I have the feeling that things would get even more stable if we
switched from storing the inverse to storing a (pivoted lu
decomposition). That would even allow to switch the indices of well
variables without getting results deviations as we have seen in the
past.

Of course DUNE should do the same and does not do that currently for
the LU solvers. But as we have our own...

**Comment by totto82** [comment]
I made a test and found a bug. The old code returned 1/det instead of det. Doesn't change the results since the determinant is not used at the moment.

**Comment by totto82** [comment]
jenkins build this opm-data=244 please

**Comment by atgeirr** [comment]
> I made a test and found a bug.

Good!

**Comment by totto82** [comment]
jenkins build this update_data please

File: opm/autodiff/BlackoilModelEbos.hpp (hunk 1 above)
Comment (totto82) [reply opm/autodiff/BlackoilModelEbos.hpp]: Yes, but currently we don't test flow_ebos_solvent. So this line is not the reason for the test failure.

**Comment by akva2** [comment]
same github issue, increased sleep to 20 secs, and created the pr for you.

**Comment by totto82** [comment]
jenkins build this opm-data=248 please

**Comment by GitPaean** [comment]
Just out of curiosity, how does this PR affect `flow_polymer`? Why the reference for `flow_polymer` needs to be updated in OPM/opm-data/#248 ?

**Comment by totto82** [comment]
Flow_polymer also gives 4x4 local matrices and is effected by this PR, but the change in solution is minor. (The test for polymer did not fail)

**Comment by totto82** [comment]
jenkins build this opm-data=248 please

**Comment by atgeirr** [comment]
Thanks for the patch, merging as all is green.

**Comment by GitPaean** [comment]
@totto82 , could you please point to me where this `MatrixBlock` gets used? I could not find it.

**Comment by totto82** [comment]
The invertMatrix is a specialization of invert for 4x4 blocks (polymer and solvent) used in ilu.hh in dune-istl.
